content/NNN-one-element-tuple-syntax.md

```diff
@@ -0,0 +1,293 @@
+---
+layout: sip
+title: Trailing Comma Syntax for Tuple Types and Values
+stage: pre-sip
+status: draft
+---
+
+# SIP-NN: Trailing Comma Syntax for Tuple Types and Values
+
+**By: [Ruslan Shevchenko]**
+
+## History
+
+| Date          | Version       |
+|---------------|---------------|
+| Nov 30th 2025 | Expanded grammar specification and SIP-27 interaction |
+| Nov 29th 2025 | Initial Draft |
+
+## Summary
+
+This proposal introduces trailing comma syntax for tuples, allowing `(A,)` to represent a single-element tuple type and `(a,)` to represent a single-element tuple value. This addresses the long-standing ambiguity where `(A)` is parsed as a parenthesized expression rather than a tuple, making it impossible to express single-element tuples using the concise parenthesized notation.
+
+The proposal also allows `(,)` for empty tuples (equivalent to `EmptyTuple`) and trailing commas in multi-element tuples like `(A, B,)`, providing consistency with other Scala constructs that already support trailing commas.
```

Comment:
As far as I know, no Scala construct supports trailing commas without a newline following them.

That also means that `(a,)` is already valid syntax, if you have a newline after the value:

```scala
scala> (
     | 1,
     | )
val res0: Int = 1
```

Seems a bit sketchy to have `(1,)` and `(1,\n)` mean different things. I personally find it very common to find `(a, b)` tuple literals growing large enough that I spread them over multiple lines, and I expect it would be common for single-element tuples as well. With this proposal, we would see a behavior change when a tuple like `(a,)` grew big enough that the opening and closing parens get moved to their own line

Reply:
Oops, then the maximum we can do is to fix the Tuple arity.